metadata-ingestion/src/datahub/sdk/lineage_client.py

```diff
@@ -220,16 +217,152 @@ def add_dataset_transform_lineage(
+
+    def add_dataset_lineage_from_sql(
+        self,
+        *,
+        query_text: str,
+        platform: str,
+        platform_instance: Optional[str] = None,
+        env: str = "PROD",
+        default_db: Optional[str] = None,
+        default_schema: Optional[str] = None,
+        schema_aware: bool = True,
+    ) -> None:
+        """Add lineage by parsing a SQL query."""
+        from datahub.sql_parsing.sqlglot_lineage import create_lineage_sql_parsed_result
+
+        # Parse the SQL query to extract lineage information
+        parsed_result = create_lineage_sql_parsed_result(
+            query=query_text,
+            default_db=default_db or "",
+            default_schema=default_schema or "",
+            platform=platform,
+            platform_instance=platform_instance,
+            env=env,
+            graph=self._client._graph,
+            schema_aware=schema_aware,
+        )
+
+        # Validate parsing result
+        if not parsed_result:
+            raise ValueError("Failed to parse SQL query: Unable to parse")
+        if parsed_result.debug_info.error:
+            raise ValueError(
+                f"Failed to parse SQL query: {parsed_result.debug_info.error}"
+            )
+        if not parsed_result.out_tables:
+            raise ValueError(
+                "No output tables found in the query. Cannot establish lineage."
+            )
+
+        # Use the first output table as the downstream
+        downstream_urn = parsed_result.out_tables[0]
+
+        # Process all upstream tables found in the query
+        for i, upstream_table in enumerate(parsed_result.in_tables):
+            # Skip self-lineage
+            if upstream_table == downstream_urn:
+                continue
+
+            # Extract column-level lineage for this specific upstream table
+            column_mapping = {}
+            if parsed_result.column_lineage:
+                for col_lineage in parsed_result.column_lineage:
+                    if not (col_lineage.downstream and col_lineage.downstream.column):
+                        continue
+
+                    # Filter upstreams to only include columns from current upstream table
+                    upstream_cols = [
+                        ref.column
+                        for ref in col_lineage.upstreams
+                        if ref.table == upstream_table and ref.column
+                    ]
+
+                    if upstream_cols:
+                        column_mapping[col_lineage.downstream.column] = upstream_cols
+
+            # Add lineage, including query text only for the first upstream
+            self.add_dataset_transform_lineage(
+                upstream=upstream_table,
+                downstream=downstream_urn,
+                column_lineage=column_mapping or None,
+                query_text=query_text if i == 0 else None,
+            )  # TODO: this throws error when table does not exist
+
+    def add_datajob_lineage(
```

Comment:
@hsheth2 let me know if this makes sense -- this just accepts upstream and downstream, and figure out the types internally

Reply:
imo we should require a datajob urn, and then upstreams can be dataset/datajob and downstream can be datasets